### ISSUE TITLE
Memory safety issues from mutable aliasing

### ISSUE THREAD

**Issue by kmcallister** [PistonDevelopers/image#362]
`png/decoder.rs` contains this snippet copy-pasted a few times:

```rust
// Unsafe copy create two views into the vector
// This is unproblematic since it is only locally to this function and a &[u8]
let data = unsafe {
    let view: &mut [u8] = mem::transmute_copy(&buf);
    &view[..entries]
};
```

Are you *sure* this is safe? `unsafe` blocks [must avoid undefined behavior](http://doc.rust-lang.org/reference.html#behavior-considered-undefined) which includes aliasing of `&mut`, full stop. LLVM can optimize under the assumption that such aliasing does not happen.

The `MutPixels` iterator may have a similar issue. Using `transmute_copy` to circumvent aliasing rules seems to be a common pattern in the library.

I think it *would* be safe to split a slice into mutually-disjoint immutable slices, then transmute them to `&mut` slices. And you can provide a safe interface to this -- I've seen it done for data parallelism, in particular. I don't know if that covers all of your use cases, however.

**Comment by nwin** [comment]
> Are you sure this is safe?
Pretty much. If it is a good idea? No.

> The `MutPixels` iterator may have a similar issue. 
I hope not, as it is as safe as `std::slice::ChunksMut`.

> Using `transmute_copy` to circumvent aliasing rules seems to be a common pattern in the library.
I think it is only used twice during image expansion.

**Comment by kmcallister** [comment]
> Pretty much.

Uh, could you elaborate on that? The Rust Reference is not enough of a precise spec to really go language-lawyering. But it strongly implies that `mem::transmute_copy` on a `&mut` pointer would be [undefined behavior](http://blog.llvm.org/2011/05/what-every-c-programmer-should-know.html) from the get-go.

How do you *know* it's not going to make an unsound optimization as a result?

**Comment by nwin** [comment]
I don’t know what LLVM could do but, but what could happen in the worst case? I can’t imagine anything that would lead to memory corruption. Not more than `std::slice::split_at_mut` would do at least. But don’t get me wrong I’m not arguing in favor of this implementation it’s just a matter of priorities and time.

**Comment by tomaka** [comment]
Maybe LLVM could reverse the loop for example?

**Comment by nwin** [comment]
Yes, but that would not be exploitable and is catchable by tests (we need to add tests!!!). The only non-deterministic behavior would happen if LLVM were to parallelize the loop. Still, I agree that is unacceptable for 1.0.

**Comment by kmcallister** [comment]
> I don’t know what LLVM could do but, but what could happen in the worst case? I can’t imagine anything that would lead to (exploitable) memory corruption.

Argument by lack of imagination isn't worth much in computer security.  Undefined behavior is a security issue *by definition* because the program's behavior could be *anything*.  Just because you or I can't figure out a way to exploit it doesn't guarantee nobody will.  I always link people to [this article](http://phrack.org/issues/57/8.html). Or you can read about some of the Linux vulns that looked benign at source level but become exploitable due to compiler optimization.

A game engine and a web browser have very different security requirements. In Servo, we can't rely on "what could happen in the worst case?" logic. Maybe this means Servo can't use Piston's image library, at least in the long run. But I'd like to help y'all make the library more secure if you are open to it.

In that case maybe I should prepare a PR to replace some of the dodgy transmutes.

**Comment by nwin** [comment]
> But I'd like to help y'all make the library more secure if you are open to it.

We are always open to it.

> In that case maybe I should prepare a PR to replace some of the dodgy transmutes.

You’re welcome. But this it a easy thing, should be something like this (can’t test because rustup broke things… :disappointed: ):

    #[inline(always)]
    fn expand_packed<F>(buf: &mut [u8], channels: usize, bit_depth: u8, func: F)
    where F: Fn(u8, &mut[u8]) {
        let entries = buf.len()/(8/bit_depth as usize);
        let mask = ((1u16 << bit_depth) - 1) as u8;
        let i =
            (0..entries)
            .rev() // Reverse iterator
            .flat_map(|idx|
                // This has to be reversed to
                (0 .. 8).step_by(bit_depth)
                .zip(iter::iterate(
                    idx, |idx| idx
                )
            ));
        let channels = channels as isize;
        let j = (buf.len() as isize - channels..-(channels)).step_by(-channels);
        //let j = (0..buf.len()).step_by(channels).rev() // ideal solution;
        for ((shift, i), j) in i.zip(j) {
            let pixel = (buf[i] & (mask << shift)) >> shift;
            func(pixel, &mut buf[j as usize..(j + channels) as usize])
        }
    }

    fn expand_palette(buf: &mut[u8], palette: &[(u8, u8, u8)],
                      entries: usize, bit_depth: u8, trns: Option<&[u8]>) {
        if let Some(trns) = trns {
            expand_packed(buf, 4, bit_depth, |i, chunk| {
                let ((r, g, b), a)) = (
                    palette[i as usize],
                    *trns.get(i as usize).unwrap_or(&0xFF)
                )
                chunk[0] = r;
                chunk[1] = g;
                chunk[2] = b;
                chunk[3] = a;
            })
        } else {
            expand_packed(buf, 3, bit_depth, |i, chunk| {
                let (r, g, b) = palette[i as usize];
                chunk[0] = r;
                chunk[1] = g;
                chunk[2] = b;
            })
        }
    }

**Comment by bvssvni** [comment]
@kmcallister thanks for bringing this up! @nwin is doing the heavy lifting for this library, and we haven't reviewed the code carefully despite the widespread use of it. This would probably not been caught otherwise.

**Comment by nwin** [comment]
@kmcallister I’ve a fix #366
